### PR TITLE
Flatten observations in the policy if needed

### PR DESCRIPTION
**Pull request by avnishn**
The current behavior of garage is for samplers,
algorithms, and users to flatten their observations
before feeding them into policies. This is because
garage policies flatten their observation spaces
and expect observations from that flattened observation
space.

The proposed change is to flatten observations inside of
the policies if needed. Observations should be flattened
only if they arent numpy arrays, because numpy arrays are
the type of observation that would be returned by a flattened
observation box space.

This makes ddpg-fetch-her work, however some additional changes
were needed to make it work. For example, the timelimit terminated
fetch environment only inserts timelimit truncated signals
if the environment has reached a timelimit. This doesn't work
when considering trajectory batches, which require that the
same number of keys are in every observation in a trajectory.
To get around this, if there is no time limit termination,
a time limit termination False signal is inserted.

Something that I need help on is how to document observations when they aren't 
numpy arrays. For example, if they come from dict spaces, they are dictionary observations.

### CODE REVIEW

**Comment by maliesa96** [comment]
CNN policies expected flattened observations and unflattened them if necessary. Was this feature removed?

**Comment by ryanjulian** [comment]
> For example, the timelimit terminated
> fetch environment only inserts timelimit truncated signals
> if the environment has reached a timelimit.

Can you split this fix into a different PR?

File: src/garage/np/policies/policy_utils.py [hunk 1]
```diff
@@ -0,0 +1,40 @@
+"""Module for utility functions commonly used by policies."""
```
Comment (ryanjulian) [inline src/garage/np/policies/policy_utils.py:1]: please do not introduce new `_utils` modules. This should go in garage.np and src/garage/np/_functions.py

File: src/garage/np/policies/policy_utils.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+        observation_space(akro.Space): The observation space that
+            observations comes from.
+        observations(observation_space.dtype): The observations to be flattened
+            if necessary.
+
+    Returns:
+        np.ndarray: Flattened observations.
+    """
+    if isinstance(observations, list):
+        if isinstance(observations[0], np.ndarray):
```
Comment (ryanjulian) [inline src/garage/np/policies/policy_utils.py:36]: does this really tell me whether it's flat or not? what if the ndarray has more than 2 dimensions?

**Comment by avnishn** [comment]
@maliesa96 no the feature was not removed. The issue is that there are observations that come from various spaces (box, dict, etc), but garage was really built around box spaces as they are the most common. The change in this pr will not change the current behavior of CNN policies.

File: src/garage/np/policies/policy_utils.py (hunk 2 above)
Comment (avnishn) [reply src/garage/np/policies/policy_utils.py:36]: Hmm. If the ndarray has more than 2 dimensions, we don't flatten it. There are still experiments that use 2d ndarray observations. We wouldn't want to flatten those because then they wouldn't match the flattened observation space. The reason why I added this if statement is because though it is documented that a user should pass numpy array observations, there were tests that passed lists of numpy arrays to the function `get_actions`

File: src/garage/tf/policies/categorical_cnn_policy.py [hunk 3]
```diff
@@ -152,6 +153,8 @@ def get_action(self, observation):
             dict(numpy.ndarray): Distribution parameters.
 
         """
+        observation = flatten_if_unflattened(self.observation_space,
```
Comment (yeukfu) [inline src/garage/tf/policies/categorical_cnn_policy.py:156]: Does the doctring of args need to change?

**Comment by krzentner** [comment]
> The current behavior of garage is for samplers,
> algorithms, and users to flatten their observations
> before feeding them into policies. This is because
> garage policies flatten their observation spaces
> and expect observations from that flattened observation
> space.

Honestly, this isn't even consistently true. As far as I can tell, the current behavior is that `get_action` and `get_actions` are *usually* given unflattened observations, and in PyTorch `forward` is given flattened observations. In TensorFlow we construct placeholders for unflattened observations, but don't feed them correctly except for images.

File: src/garage/_functions.py [hunk 4]
```diff
@@ -159,3 +159,43 @@ def log_performance(itr, batch, discount, prefix='Evaluation'):
...
+    Args:
+        observation_space(akro.Space): The observation space that
+            observation comes from.
+        observation(observation_space.dtype): The observations to be flattened
+            if necessary.
+
+    Returns:
+        np.ndarray: Flattened observations.
+    """
+    if not isinstance(observation, np.ndarray):
```
Comment (krzentner) [inline src/garage/_functions.py:176]: I think this should also check that `len(observation.shape) == 1`

File: src/garage/_functions.py [hunk 5]
```diff
@@ -159,3 +159,43 @@ def log_performance(itr, batch, discount, prefix='Evaluation'):
...
+            if necessary.
+
+    Returns:
+        np.ndarray: Flattened observations.
+    """
+    # if a list of observations is passed, then
+    # don't flatten them if the individual observations
+    # are numpy arrays
+    if isinstance(observations, list):
+        if isinstance(observations[0], np.ndarray):
```
Comment (krzentner) [inline src/garage/_functions.py:197]: Same here.

File: src/garage/_functions.py [hunk 6]
```diff
@@ -159,3 +159,43 @@ def log_performance(itr, batch, discount, prefix='Evaluation'):
...
+            return observations
+    if not isinstance(observations, np.ndarray):
```
Comment (krzentner) [inline src/garage/_functions.py:199]: Should probably check that `len(observation.shape) == 2`.

File: src/garage/tf/policies/categorical_cnn_policy.py (hunk 3 above)
Comment (krzentner) [inline src/garage/tf/policies/categorical_cnn_policy.py:156]: This is really confusing. Basically, the code here looks like it flattens images, but the code for doing it won't work correctly (because unflattened images are still `np.ndarray`s), and then the next line requires unflattened images. I think the right thing to do here is just to not flatten the observations.

File: tests/garage/sampler/test_utils.py [hunk 7]
```diff
@@ -23,11 +23,6 @@ def test_max_path_length(self):
         assert path['agent_infos']['dummy'].shape[0] == 3
         assert path['env_infos']['dummy'].shape[0] == 3
 
-    def test_does_flatten(self):
```
Comment (krzentner) [inline tests/garage/sampler/test_utils.py:26]: Wait, shouldn't this test still pass? Or maybe it's named wrong.

File: src/garage/_functions.py [hunk 8]
```diff
@@ -159,3 +159,43 @@ def log_performance(itr, batch, discount, prefix='Evaluation'):
             tabular.record('SuccessRate', np.mean(success))
 
     return undiscounted_returns
+
+
+def flatten_if_unflattened(observation_space, observation):
+    """Flattens observation to a numpy.ndarray if it is not.
```
Comment (haydenshively) [inline src/garage/_functions.py:165]: This isn't very clear, at least to me. In numpy-speak, flattened would mean `ndarray.ndim == 1`, but it looks like all this function does is ensure `isinstance(observation, np.ndarray)`. Basically, the `ndim` of the returned array is unconstrained, so "flattened" doesn't really mean what it usually means.

File: tests/garage/sampler/test_utils.py (hunk 7 above)
Comment (avnishn) [reply tests/garage/sampler/test_utils.py:26]: no this test checks to see that the sampler is flattening observations. After a discussion we had, the conclusion we came to was that the sampler shouldn't be doing any flattening or unflattening.

**Comment by avnishn** [comment]
perhaps flatten is the wrong word to use here. As @haydenshively pointed out, 

> flattened would mean ndarray.ndim == 1, but it looks like all this function does is ensure isinstance(observation, np.ndarray)

Thats not the goal of this change. The goal is to verify that the observations are numpy arrays before feeding them into the policies.

Also @krzentner to answer your question, yes this does fix the ddpg-fetchreach-her example.

File: src/garage/_functions.py [hunk 9]
```diff
@@ -159,3 +159,43 @@ def log_performance(itr, batch, discount, prefix='Evaluation'):
...
+    Args:
+        observation_space(akro.Space): The observation space that
+            observations comes from.
+        observations(observation_space.dtype): The observations to be flattened
+            if necessary.
+
+    Returns:
+        np.ndarray: Flattened observations.
+    """
+    # if a list of observations is passed, then
```
Comment (ryanjulian) [inline src/garage/_functions.py:193]: why not just make this one function which checks the input type?

**Comment by avnishn** [comment]
@mergifyio rebase

File: src/garage/tf/policies/categorical_cnn_policy.py [hunk 10]
```diff
@@ -152,6 +153,7 @@ def get_action(self, observation):
             dict(numpy.ndarray): Distribution parameters.
 
         """
+        observation = convert_to_numpy(self.observation_space, observation)
```
Comment (krzentner) [inline src/garage/tf/policies/categorical_cnn_policy.py:156]: These really shouldn't be necessary for this policy, since if your input isn't a valid image, flattening it certainly won't make it into one.

File: src/garage/_functions.py [hunk 11]
```diff
@@ -163,3 +163,43 @@ def log_performance(itr, batch, discount, prefix='Evaluation'):
...
+
+    Returns:
+        np.ndarray: Flattened observations.
+    """
+    if not isinstance(observation, np.ndarray):
+        observation = (observation_space.flatten(observation))
+    return observation
+
+
+def convert_n_to_numpy(observation_space, observations):
```
Comment (ryanjulian) [inline src/garage/_functions.py:185]: move these to garage.np

File: src/garage/tf/policies/continuous_mlp_policy.py [hunk 12]
```diff
@@ -122,9 +123,9 @@ def get_action(self, observation):
             dict: Empty dict since this policy does not model a distribution.
 
         """
-        action = self._f_prob([observation])
-        action = self.action_space.unflatten(action)
-        return action, dict()
+        actions, agent_infos = self.get_actions(np.array([observation]))
```
Comment (ryanjulian) [inline src/garage/tf/policies/continuous_mlp_policy.py:126]: i thought you already guaranteed the observation is an np.ndarray?

File: src/garage/tf/policies/discrete_qf_derived_policy.py [hunk 13]
```diff
@@ -50,10 +50,8 @@ def get_action(self, observation):
                 len(self.env_spec.observation_space.shape):
             observation = self.env_spec.observation_space.unflatten(
                 observation)
-        q_vals = self._f_qval([observation])
-        opt_action = np.argmax(q_vals)
-
-        return opt_action, dict()
+        opt_actions, agent_infos = self.get_actions(np.array([observation]))
```
Comment (ryanjulian) [inline src/garage/tf/policies/discrete_qf_derived_policy.py:53]: why always the cast to np.ndarray?

File: tests/garage/sampler/test_utils.py (hunk 7 above)
Comment (ryanjulian) [reply tests/garage/sampler/test_utils.py:26]: perhaps you should just modify it to `test_does_not_flatten` instead

File: src/garage/tf/policies/discrete_qf_derived_policy.py (hunk 13 above)
Comment (avnishn) [reply src/garage/tf/policies/discrete_qf_derived_policy.py:53]: The method `get_actions` asks for a numpy array. I thought I would stay true to the documentation where possible.

I'll probably have to update the docs to allow for the union of all the different types of akro observations.

File: src/garage/tf/policies/continuous_mlp_policy.py (hunk 12 above)
Comment (avnishn) [reply src/garage/tf/policies/continuous_mlp_policy.py:126]: I will need to update the documentation to support the different observation types of akro.

File: src/garage/tf/policies/discrete_qf_derived_policy.py (hunk 13 above)
Comment (ryanjulian) [reply src/garage/tf/policies/discrete_qf_derived_policy.py:58]: ...but why would observation not be a numpy array after flattening? wasn't that the point?

File: tests/garage/replay_buffer/test_her_replay_buffer.py [hunk 14]
```diff
@@ -11,7 +12,8 @@
 class TestHerReplayBuffer:
 
     def setup_method(self):
-        self.env = DummyDictEnv()
+        self.env = GarageEnv(DummyDictEnv())
+        # import ipdb; ipdb.set_trace()
```
Comment (ryanjulian) [inline tests/garage/replay_buffer/test_her_replay_buffer.py:16]: remove

File: CHANGELOG.md [hunk 15]
```diff
@@ -211,7 +211,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - `garage.np.OffPolicyRLAlgorithm` ([#1552](https://github.com/rlworkgroup/garage/pull/1552))
 
 ### Fixed
-- Bug where `GarageEnv` did not pickle ([#1029](https://github.com/rlworkgroup/garage/pull/1029))
+- Bug where `GymEnv` did not pickle ([#1029](https://github.com/rlworkgroup/garage/pull/1029))
```
Comment (yeukfu) [inline CHANGELOG.md:214]: I think we doesn't need to change this name as it is in a log.

File: docs/user/implement_algo.md [hunk 16]
```diff
@@ -84,13 +84,21 @@ We'll also want an environment to test our algorithm with.
 
 ```py
 from garage import wrap_experiment
+<<<<<<< HEAD
```
Comment (yeukfu) [inline docs/user/implement_algo.md:87]: should remove?

File: docs/user/implement_algo.md [hunk 17]
```diff
@@ -84,13 +84,21 @@ We'll also want an environment to test our algorithm with.
...
 from garage.envs import PointEnv
+=======
+from garage.envs import PointEnv, GymEnv
+>>>>>>> a1877525... Flatten observations if necessary
 from garage.experiment import LocalRunner
 
 @wrap_experiment
 def debug_my_algorithm(ctxt):
     runner = LocalRunner(ctxt)
+<<<<<<< HEAD
```
Comment (yeukfu) [inline docs/user/implement_algo.md:97]: should remove?

File: docs/user/meta_multi_task_rl_exp.md [hunk 18]
```diff
@@ -50,7 +50,11 @@ def te_ppo_ml1_push(ctxt, seed, n_epochs, batch_size_per_task):
 
     """
     set_seed(seed)
+<<<<<<< HEAD
```
Comment (yeukfu) [inline docs/user/meta_multi_task_rl_exp.md:53]: should remove?

File: docs/user/training_a_policy.md [hunk 19]
```diff
@@ -47,7 +47,11 @@ like [this](implement_env). In this example, we choose `CartPole-V1`
 environment.
 
 ```py
+<<<<<<< HEAD
```
Comment (yeukfu) [inline docs/user/training_a_policy.md:50]: should remove?

File: docs/user/use_pretrained_network_to_start_new_experiment.md [hunk 20]
```diff
@@ -75,6 +75,7 @@ import gym
 
 from garage import wrap_experiment
 from garage.envs import GymEnv
+<<<<<<< HEAD
```
Comment (yeukfu) [inline docs/user/use_pretrained_network_to_start_new_experiment.md:78]: the same as above

File: examples/jupyter/custom_env.ipynb [hunk 21]
```diff
@@ -359,7 +359,11 @@
     "        return obs\n",
     "      \n",
     "env = NpWrapper(env)\n",
+<<<<<<< HEAD
```
Comment (yeukfu) [inline examples/jupyter/custom_env.ipynb:362]: the same as above

**Comment by yeukfu** [comment]
@avnishn Could you please review my comments?

**Comment by avnishn** [comment]
Yup sorry about that I didn't catch it in the tests thanks! Will fix rn.

**Comment by ryanjulian** [comment]
@Mergifyio backport release-2020.06
